File: Data Structures & Algorithms/search-for-word/submission-2.py
class Solution:
    def exist(self, board: List[List[str]], word: str) -> bool:
        def dfs(i, j, word, seen):
            if not word:
                return True

            if i < 0 or i >= len(board):
                return False
            
            if j < 0 or j >= len(board[0]):
                return False
            
            if (i, j) in seen:
                return False
            seen.add((i,j))

            if board[i][j] != word[0]:
                return False
            
            
            if dfs(i+1, j, word[1:], seen):
                return True
            if dfs(i-1, j, word[1:], seen):
                return True
            if dfs(i, j+1, word[1:], seen):
                return True
            if dfs(i, j-1, word[1:], seen):
                return True

            return False


        for i in range(len(board)):
            for j in range(len(board[0])):
                if dfs(i, j, word, set()):
                    return True
        
        return False

File: Data Structures & Algorithms/max-area-of-island/submission-2.cpp
class Solution {
public:
    int dfs(int i, int j, vector<vector<int>>& grid, vector<vector<bool>>& seen) {
        int rows = grid.size();
        int cols = grid[0].size();
        int size{};

        if (i < 0 || j < 0 || i >= rows || j >= cols) {
            return size;
        }
        if (grid[i][j] == 0) {
            return size;
        }
        if (seen[i][j] == true) {
            return size;
        }
        seen[i][j] = true;
        ++size;

        size += dfs(i+1, j, grid, seen);
        size += dfs(i-1, j, grid, seen);
        size += dfs(i, j+1, grid, seen);
        size += dfs(i, j-1, grid, seen);
        
        return size;
    }

    int maxAreaOfIsland(vector<vector<int>>& grid) {
        int rows = grid.size();
        int cols = grid[0].size();

        std::vector<std::vector<bool>> seen(rows, std::vector<bool>(cols, false));

        int max_size{};
        for (int i{}; i != rows; ++i) {
            for (int j{}; j != cols; ++j) {
                max_size = std::max(dfs(i, j, grid, seen), max_size);
            }
        }

        return max_size;
    }
};
